clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,179 @@
+# Integration Header
+
+We propose an Integration Header for two purposes:
+
+* to allow a non-SYCL-aware compiler to do host compilations and
+* as a mechanism permitting the SYCL run-time library to use the information provided for its needs
+
+Consider the following partial code:
+
+```
+  sycl::accessor<char, 1, sycl::access::mode::read> acc1;
+  int i = 13;
+  sycl::sampler smplr;
+  struct S {
+    char c;
+    int i;
+  } test_s;
+  test_s.c = 14;
+  kernel_single_task<class first_kernel>([=]() {
+    if (i == 13 && test_s.c == 14) {
+
+      acc1.use();
+      smplr.use();
+    }
+  });
+```
+
+Here the lambda within the 'kernel_single_task' construct needs to be executed on the device.  To do this, a function object representing the kernel is generated.   The function call operator of this object has the contents of the kernel invocation.  The function object generated looks like:
+
+```
+struct FuncObj {
+  int i;
+  struct S test_s;
+  sycl::accessor acc1;
+  sycl::sampler smplr;
+
+  () {  // Function call operator
+    if (i == 13 && test_s.c == 14) {
+      acc1.use();
+      smplr.use();
+    }
+  }
+};
+```
+
+The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.
+
+The device compiler transforms this into (pseudo-code):
+
+```
+    spir_kernel void caller(
+       int i,
+       struct S test_s,
+       __global int* accData, // arg1 of accessor init function
+       range<1> accR1,        // arg2 of accessor init function
+       range<1> accR2,        // arg3 of accessor init function
+       id<1> accId,           // arg4 of accessor init function
+       sampler_t smpData      // arg1 of sampler init function
+    )
+    {
+        // Local capture object
+        struct FuncObj local;
+
+        // Reassemble capture object from parts
+        local.i = i;
+        local.s = s;
+        // Call acc1 accessor's init function
+        sycl::accessor::init(&local.acc1, accData, accR1, accR2, accId);
+        // Call smplr sampler's init function
+        sycl::accessor::init(&local.smplr, smpData);
+
+        // Call the kernel body
+        callee(&local);
+    }
+
+    spir_func void callee(struct FuncObj* this)
+    {
+        // body of the kernel invocation
+    }
+```
```

Comment:
This exposes our SPIR bias. I'm not sure this is helpful either. From the integration header/footer perspective, declarations are all that really matter. I think what we want to show is how a call to a kernel invocation function is translated by a non-SYCL-aware host compiler into a call into the SYCL run-time library with references to the synthesized declarations provided by the integration header.